Utils/Services/ImagePickerService.swift

```diff
@@ -0,0 +1,160 @@
+//
+//  ImagePickerService.swift
+//  Utils
+//
+//  Created by Nahuel Gladstein on 6/13/17.
+//  Copyright © 2017 Wolox. All rights reserved.
+//
+
+import ReactiveSwift
+import UIKit
+import AVFoundation
+import Photos
+import Result
+
+public enum ImagePickerServiceError: Error {
+    
+    case sourceTypeNotAvailable
+    
+}
+
+public protocol ImagePickerServiceType {
+    /**
+     Observe imageSignal to get the UIImage selected by the user
+     */
+    var imageSignal: Signal<UIImage, ImagePickerServiceError> { get }
+    
+    /**
+     Presents the picker to the user so it can take or select a picture. If the user didn't give permission to the app to use the source type selected
+        a prompt asking it will be shown.
+     - parameter source: Source type for the picker to show. Can be .camera or .photoLibrary.
+     - parameter onPermissionNotGranted: Block called if the user denies permission. If the user gives permission the camera will be shown.
+     */
+    func presentImagePickerController(for source: UIImagePickerControllerSourceType, _ onPermissionNotGranted: @escaping (Void) -> Void)
+    
+    /**
+     Tells if the device has a camera.
+     */
+    var cameraIsAvailable: Bool { get }
+}
+
+@objc
+public final class ImagePickerService: NSObject, ImagePickerServiceType {
+    
+    public let imageSignal: Signal<UIImage, ImagePickerServiceError>
+    fileprivate let _imageObserver: Signal<UIImage, ImagePickerServiceError>.Observer
+    
+    fileprivate weak var _viewController: UIViewController?
+    
+    init(viewController: UIViewController) {
+        _viewController = viewController
+        (imageSignal, _imageObserver) = Signal<UIImage, ImagePickerServiceError>.pipe()
+    }
+    
+    public func presentImagePickerController(for source: UIImagePickerControllerSourceType, _ onPermissionNotGranted: @escaping (Void) -> Void) {
+        switch source {
+        case .camera:
+            presentCameraTypeImagePickerController(onPermissionNotGranted)
+        case .photoLibrary:
+            presentGalleryTypeImagePickerController(onPermissionNotGranted)
+        default:
+            break
+        }
+    }
+    
+    public var cameraIsAvailable: Bool {
+        return UIImagePickerController.isSourceTypeAvailable(.camera)
+    }
+    
+    deinit {
+        _imageObserver.sendCompleted()
+    }
+}
+
+fileprivate extension ImagePickerService {
+    
+    fileprivate func presentCameraTypeImagePickerController(_ onPermissionNotGranted: @escaping (Void) -> Void) {
```

Comment:
I think I would just leave the function like this:

```
public presentImagePickerController(for source: UIImagePickerControllerSourceType, _ onPermissionNotGranted: @escaping (Void) -> Void) {
    source.isPermitted().startWithResults {
        switch $0 {
        case .success(let permitted):
            if permitted { self.presentImagePickerController(source) }
            else { onPermissionNotGranted() }
        case .failure(let error): _imageObserver.send(error: error)
        }
    }
}
```
and have the source type have:
```
fileprivate extension UIImagePickerControllerSourceType {
    fileprivate func isPermitted() -> SignalProducer<Bool, ImagePickerServiceError> {
        switch self {
        case .camera:
            guard cameraIsAvailable else { return SignalProducer(error: . sourceTypeNotAvailable) }
            return hasCameraPermission()
        case .photoLibrary: return hasPhotosPermission()
        default: return .empty
        }
    }
}
```